### PR TITLE
test/e2e/framework/util.go:move DsFromManifest to test/e2e/framework/manifest , and rename it to GetDaemonSetFromURL

### PR DESCRIPTION
**Pull request by tanjunchen**
**What type of PR is this?**
  
  /kind cleanup
  /priority backlog

**What this PR does / why we need it**:

moving DsFromManifest to test/e2e/framework/manifest , And I think it is better to rename it  GetDaemonSetFromURL.

**Which issue(s) this PR fixes**:

ref:https://github.com/kubernetes/kubernetes/issues/77095

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
None
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs
None
```

### CODE REVIEW

File: test/e2e/scheduling/nvidia-gpus.go [hunk 1]
```diff
@@ -35,6 +35,7 @@ import (
 	"k8s.io/kubernetes/test/e2e/framework/providers/gce"
 	e2eresource "k8s.io/kubernetes/test/e2e/framework/resource"
 	e2eskipper "k8s.io/kubernetes/test/e2e/framework/skipper"
+	manifestutils "k8s.io/kubernetes/test/e2e/manifest"
```
Comment (oomichi) [inline test/e2e/scheduling/nvidia-gpus.go]: How about `e2emanifest` here like the above?

File: test/e2e_node/gpu_device_plugin_test.go [hunk 2]
```diff
@@ -29,6 +29,7 @@ import (
 	"k8s.io/kubernetes/test/e2e/framework"
 	"k8s.io/kubernetes/test/e2e/framework/gpu"
 	"k8s.io/kubernetes/test/e2e/framework/metrics"
+	manifestutils "k8s.io/kubernetes/test/e2e/manifest"
```
Comment (oomichi) [inline test/e2e_node/gpu_device_plugin_test.go]: ditto

File: test/e2e_node/image_list.go [hunk 3]
```diff
@@ -32,6 +32,7 @@ import (
 	"k8s.io/kubernetes/test/e2e/framework"
 	"k8s.io/kubernetes/test/e2e/framework/gpu"
 	"k8s.io/kubernetes/test/e2e/framework/testfiles"
+	manifestutils "k8s.io/kubernetes/test/e2e/manifest"
```
Comment (oomichi) [inline test/e2e_node/image_list.go]: ditto

**Review by oomichi — COMMENTED** [review]
Nice cleanup, thanks for doing this.
just few comments.

File: test/e2e/scheduling/nvidia-gpus.go (hunk 1 above)
Comment (tanjunchen) [reply test/e2e/scheduling/nvidia-gpus.go]: manifestutils is also ok , but i will update it . thanks

Comment (tanjunchen) [reply test/e2e/scheduling/nvidia-gpus.go]: done

**Comment by tanjunchen** [comment]
/cc @neolit123 
😄  again

File: test/e2e/manifest/manifest.go [hunk 4]
```diff
@@ -35,7 +40,7 @@ func PodFromManifest(filename string) (*v1.Pod, error) {
 		return nil, err
 	}
 
-	json, err := utilyaml.ToJSON(data)
+	json, err := yamlutil.ToJSON(data)
```
Comment (neolit123) [inline test/e2e/manifest/manifest.go]: was this rename needed?

Comment (tanjunchen) [reply test/e2e/manifest/manifest.go]: i prefer to yamlutil , i think helpers usually ends with util .
:smile:

Comment (tanjunchen) [reply test/e2e/manifest/manifest.go]: util/yaml  -> utilyaml 
I know the reason , revert this change, make it much clear for this pr.

Comment (neolit123) [reply test/e2e/manifest/manifest.go]: yes, preferably the PR should do one understandable change. if there are multiple changes please add them in separate commits and explain all commits in the PR description.

**Comment by neolit123** [comment]
the looks mostly OK to me, leaving the final review to @oomichi

File: test/e2e/scheduling/nvidia-gpus.go [hunk 5]
```diff
@@ -35,6 +35,7 @@ import (
 	"k8s.io/kubernetes/test/e2e/framework/providers/gce"
 	e2eresource "k8s.io/kubernetes/test/e2e/framework/resource"
 	e2eskipper "k8s.io/kubernetes/test/e2e/framework/skipper"
+	e2emanifest "k8s.io/kubernetes/test/e2e/manifest"
```
Comment (alejandrox1) [inline test/e2e/scheduling/nvidia-gpus.go]: I would vote to alias this package as something like `manifestutils` or `testmanifests` or something of the sort.
Up to now we have been using the `e2e` prefix for framework subpackages.
I.e., if it the package was in `k8s.io/kubernetes/test/e2e/framework/manifest` then that one would be `e2emanifest`.

File: test/e2e_node/gpu_device_plugin_test.go [hunk 6]
```diff
@@ -29,6 +29,7 @@ import (
 	"k8s.io/kubernetes/test/e2e/framework"
 	"k8s.io/kubernetes/test/e2e/framework/gpu"
 	"k8s.io/kubernetes/test/e2e/framework/metrics"
+	e2emanifest "k8s.io/kubernetes/test/e2e/manifest"
```
Comment (alejandrox1) [inline test/e2e_node/gpu_device_plugin_test.go]: [1]

File: test/e2e_node/image_list.go [hunk 7]
```diff
@@ -32,6 +32,7 @@ import (
 	"k8s.io/kubernetes/test/e2e/framework"
 	"k8s.io/kubernetes/test/e2e/framework/gpu"
 	"k8s.io/kubernetes/test/e2e/framework/testfiles"
+	e2emanifest "k8s.io/kubernetes/test/e2e/manifest"
```
Comment (alejandrox1) [inline test/e2e_node/image_list.go]: [1]

File: test/e2e/manifest/manifest.go [hunk 8]
```diff
@@ -105,42 +92,47 @@ func StatefulSetFromManifest(fileName, ns string) (*appsv1.StatefulSet, error) {
...
 }
 
-// DaemonSetFromManifest returns a DaemonSet from a manifest stored in fileName in the Namespace indicated by ns.
-func DaemonSetFromManifest(fileName, ns string) (*appsv1.DaemonSet, error) {
-	var ds appsv1.DaemonSet
-	data, err := testfiles.Read(fileName)
-	if err != nil {
-		return nil, err
+// DaemonSetFromManifest reads a .json/yaml file and returns the daemonset in it.
+func DaemonSetFromManifest(url string) (*appsv1.DaemonSet, error) {
```
Comment (alejandrox1) [inline test/e2e/manifest/manifest.go]: We should update the name and description of this function to indicate that the manifest is downloaded from a URL.
`<Something>FromManifest` makes me think that we are passing the json or yaml directly instead of specifying a URL.

**Review by alejandrox1 — COMMENTED** [review]
Thank you very much for looking into these functions @tanjunchen !
left some minor comments but lgtm otherwise

File: test/e2e/scheduling/nvidia-gpus.go [hunk 9]
```diff
@@ -137,7 +138,7 @@ func SetupNVIDIAGPUNode(f *framework.Framework, setupResourceGatherer bool) *fra
 
 	framework.Logf("Using %v", dsYamlURL)
 	// Creates the DaemonSet that installs Nvidia Drivers.
-	ds, err := framework.DsFromManifest(dsYamlURL)
+	ds, err := e2emanifest.DaemonSetFromManifest(dsYamlURL)
```
Comment (Huang-Wei) [inline test/e2e/scheduling/nvidia-gpus.go]: Can we rename `DaemonSetFromManifest` to `DaemonSetFromURL`?

Comment (tanjunchen) [reply test/e2e/scheduling/nvidia-gpus.go]: sure . thanks your review.

File: test/e2e/scheduling/nvidia-gpus.go (hunk 1 above)
Comment (tanjunchen) [reply test/e2e/scheduling/nvidia-gpus.go]: /cc @oomichi 
```
Up to now we have been using the e2e prefix for framework subpackages.
I.e., if it the package was in k8s.io/kubernetes/test/e2e/framework/manifest then that one would be e2emanifest
```
That sounds good. how about manifestutils? I think it is better currently.

File: test/e2e/scheduling/nvidia-gpus.go [hunk 10]
```diff
@@ -137,7 +138,7 @@ func SetupNVIDIAGPUNode(f *framework.Framework, setupResourceGatherer bool) *fra
 
 	framework.Logf("Using %v", dsYamlURL)
 	// Creates the DaemonSet that installs Nvidia Drivers.
-	ds, err := framework.DsFromManifest(dsYamlURL)
+	ds, err := manifestutils.DaemonSetFromURL(dsYamlURL)
```
Comment (tanjunchen) [inline test/e2e/scheduling/nvidia-gpus.go]: /cc @Huang-Wei 
done

File: test/e2e/scheduling/nvidia-gpus.go (hunk 9 above)
Comment (alejandrox1) [reply test/e2e/scheduling/nvidia-gpus.go]: ha! this has a better ring to it! :+1:

**Comment by alejandrox1** [comment]
From https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/89456/pull-kubernetes-bazel-build/1243737320662765568#1:build-log.txt%3A65

```
/b/f/w/test/e2e_node/gpu_device_plugin_test.go:52:13: undefined: "k8s.io/kubernetes/test/e2e/framework".DsFromManifest
```

but overall looks good.
The new names make more sense :rocket:

**Comment by tanjunchen** [comment]
> From https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/89456/pull-kubernetes-bazel-build/1243737320662765568#1:build-log.txt%3A65
> 
> ```
> /b/f/w/test/e2e_node/gpu_device_plugin_test.go:52:13: undefined: "k8s.io/kubernetes/test/e2e/framework".DsFromManifest
> ```
> 
> but overall looks good.
> The new names make more sense 🚀

updated

File: test/e2e/scheduling/nvidia-gpus.go (hunk 1 above)
Comment (oomichi) [reply test/e2e/scheduling/nvidia-gpus.go]: I don't think manifestutils is better than e2e manifest, because all e2e test subpackage is named as "e2e" + "<subpackage name>" like https://github.com/kubernetes/kubernetes/pull/89454/files
Why do you want to name this on different way?

Comment (oomichi) [reply test/e2e/scheduling/nvidia-gpus.go]: humm, I need to think this again.
It is fine to move a function from `e2e/framework/util.go` into specific e2e test package if the function is used at the test package only.
However, this makes another dependency from 2 different test packages (e2e/scheduling and e2e_node) without e2e test framework which is shared place for different test packages. That is different from the existing way, is that intentional?

/lgtm cancel

Comment (tanjunchen) [reply test/e2e/scheduling/nvidia-gpus.go]: yes , you are right. It is different from the existing way. 
But  I  think it is  good to  maintain *FromURL functions in test/e2e/manitest package . They are the same or  homologous function.

Comment (oomichi) [reply test/e2e/scheduling/nvidia-gpus.go]: Humm, I am not sure this direction is valid for us.
e2e test framework should be a single place to hold shared functions between multiple e2e test packages, I did think before reviewing this PR.
If going this PR direction, I am not sure what is e2e test framework is.
I'd like to see more opinions about this.

/hold

**Comment by Huang-Wei** [comment]
/approve
on scheduling side.

**Comment by oomichi** [comment]
Good renaming

/test pull-kubernetes-kubemark-e2e-gce-big
/hold cancel
/lgtm

**Review by alejandrox1 — COMMENTED** [review]
Thank you @tanjunchen !

/kind cleanup
/priority backlog
/approve

File: test/e2e/scheduling/nvidia-gpus.go (hunk 1 above)
Comment (alejandrox1) [reply test/e2e/scheduling/nvidia-gpus.go]: we can make an argument for moving that method outside of the e2e framework as the framework is not only mean to hold shared functions between e2e packages but it is meant to hold useful functions for e2e tests. Instantiating a daemonset from a manifest that can be download from a URL is not something to have in the e2e framework.
As is, it is good to move it out. To keep it in the framework (one of its subpackages) we should rewrite it to create any sort of Kubernetes resource based on a manifest that can be specified by a URL.

**Comment by alejandrox1** [comment]
actually...
/hold

There is one good aspect to discuss from this PR:

originally this PR moved the function `GetDaemonSetFromURL` from out of the e2e test framework.
This is the only function of this sort in the framework - one which creates a Kubernetes resource based on a manifest specified by a URL.

The framework is meant to hold shared code to make writing e2e tests easier and not all shared code is the same.
Leaving the function where it is not a good final decision - ultimately because this functionality doesn't belong in the e2e test framework.
One place where it logically belongs could be in https://github.com/kubernetes/kubernetes/tree/master/test/e2e/manifest
and even that package seems useful enough that we may want to include it in the e2e test framework as a subpackage.

Before we move on from this PR we should get some consensus about what to do with this function and functions like this one.
Because leaving it as is is not the ultimate goal.

what do you all think?

**Comment by oomichi** [comment]
@alejandrox1 
> ultimately because this functionality doesn't belong in the e2e test framework. One place where it logically belongs could be in https://github.com/kubernetes/kubernetes/tree/master/test/e2e/manifest

Could you explain more why you are thinking the above?

**Comment by tanjunchen** [comment]
ok . we should get some consensus about what to do with this function. move it to test/e2e/manifest or subpackge test/e2e/framework/manifest or stay it here.

**Comment by neolit123** [comment]
> This is the only function of this sort in the framework - one which creates a Kubernetes resource based on a manifest specified by a URL.

this seems like a good point.
so +1 to move the function outside to e2e/manifest, but no strong opinions from me.

we should note that `/test` has an issue that we have frameworks within frameworks...
as long as we can make `/test/e2e/framework` not depend on the rest of k/k that would be great progress.

**Comment by alejandrox1** [comment]
> @alejandrox1
> 
> > ultimately because this functionality doesn't belong in the e2e test framework. One place where it logically belongs could be in https://github.com/kubernetes/kubernetes/tree/master/test/e2e/manifest
> 
> Could you explain more why you are thinking the above?

Taking a look at https://github.com/kubernetes/kubernetes/blob/master/test/e2e/manifest/manifest.go
every function there deals with creating some sort of Kubernetes resource from a manifest, essentially same sort of work as the function being worked on here - additional step being that the manifest comes from a URL.

I think that based on the approach we have had of keeping the e2e framework just for e2e test configuration (dealing with config management, logging, etc) and stripping away all other utilities used in e2e tests (i.e., managing pods, nodes), having a function that creates a daemonset does not fit in with e2e framework itself.
I think it would fit in better with the e2e/manifest pkg at this point in time.

**Comment by oomichi** [comment]
@alejandrox1 
> Taking a look at https://github.com/kubernetes/kubernetes/blob/master/test/e2e/manifest/manifest.go
every function there deals with creating some sort of Kubernetes resource from a manifest, essentially same sort of work as the function being worked on here - additional step being that the manifest comes from a URL.

From the viewpoint of functionality, the above point is reasonable for me.
`test/e2e/manifest/manifest.go/` file is a good place to keep functions which returns values from specified paths(filename, and additional URL).

However at this time, the packages which import this file are only 2 places like:
```
$ grep "k8s.io/kubernetes/test/e2e/manifest" * -R
e2e/e2e.go:     "k8s.io/kubernetes/test/e2e/manifest"
e2e/framework/statefulset/rest.go:      "k8s.io/kubernetes/test/e2e/manifest"
```
e2e/e2e.go is a core part of e2e test and it should consists of multiple e2e test packages, then it is fine to import `k8s.io/kubernetes/test/e2e/manifest`.
e2e/framework/statefulset/rest.go should not import `k8s.io/kubernetes/test/e2e/manifest` and we will remove this import soon based on https://github.com/kubernetes/kubernetes/issues/74352

So there are not any dependencies to `k8s.io/kubernetes/test/e2e/manifest` from the other e2e test packages except the core part of e2e test. If we move DsFromManifest() into `k8s.io/kubernetes/test/e2e/manifest` we will have more dependencies between e2e test packages without e2e test framework. That makes me hesitate to go to the direction.

**Comment by alejandrox1** [comment]
I was going to write a very long paragraph explaining my idea but in this case a PR was easier.
But overall, what i'm thinking, `k8s.io/kubernetes/test/e2e/manifest` looks useful for e2e tests.
If we can move it to the framework, as in https://github.com/kubernetes/kubernetes/pull/90035 , then we can move things like `GetDaemonSetFromURL()` to the package and it would make sense.
 (Plus we would remove the dependency on an external package from `e2e/framework/statefulset/rest.go`).

Howeverrrr...
> If we move DsFromManifest() into k8s.io/kubernetes/test/e2e/manifest we will have more dependencies between e2e test packages without e2e test framework.

I'm don't think I completely understand what you mean there.
Could you elaborate more, please?

**Comment by oomichi** [comment]
>> If we move DsFromManifest() into k8s.io/kubernetes/test/e2e/manifest we will have more dependencies between e2e test packages without e2e test framework.
>
> I'm don't think I completely understand what you mean there.
> Could you elaborate more, please?

Sorry for my previous unclear comment.
I just did think e2e test packages (like `test/e2e/manifest`, `test/e2e/auth`, etc.) should not have dependencies each other except e2e test framework (`test/e2e/framework`).

**Comment by oomichi** [comment]
@alejandrox1 
https://github.com/kubernetes/kubernetes/pull/90035 seems good for me, thanks for doing it :-)

**Comment by tanjunchen** [comment]
/cc @alejandrox1 
/cc @oomichi 
for review.

File: test/e2e/framework/manifest/manifest.go [hunk 11]
```diff
@@ -85,3 +91,48 @@ func StatefulSetFromManifest(fileName, ns string) (*appsv1.StatefulSet, error) {
 	}
 	return &ss, nil
 }
+
+// GetDaemonSetFromURL reads from a url and returns the daemonset in it.
+func GetDaemonSetFromURL(url string) (*appsv1.DaemonSet, error) {
```
Comment (oomichi) [inline test/e2e/framework/manifest/manifest.go:96]: Why do you rename this function only?
I feel this naming way (GetFooFromBar() ) seems good for me, but this way is different from the existing functions.
That makes inconsistency at this time.
How about just moving the function without renaming now?

Comment (tanjunchen) [reply test/e2e/framework/manifest/manifest.go:96]: see https://github.com/kubernetes/kubernetes/pull/89456#discussion_r399462716

Comment (alejandrox1) [inline test/e2e/framework/manifest/manifest.go:96]: nit: can we rename it to `DaemonSetFromURL`.
For the most part, it does "Get" a daemonset from a URL but it also creates it in the last step.

**Review by alejandrox1 — COMMENTED** [review]
Small nit on the name but otherwise this looks great!
Thank you @tanjunchen for pushing this forward!

File: test/e2e/framework/manifest/manifest.go (hunk 11 above)
Comment (tanjunchen) [reply test/e2e/framework/manifest/manifest.go:96]: ok .  i will do it.

Comment (oomichi) [reply test/e2e/framework/manifest/manifest.go:96]: Yeah, my point was `Get` part.
So it is nice to rename it to `DaemonSetFromURL`.

File: test/e2e/framework/manifest/manifest.go [hunk 12]
```diff
@@ -85,3 +91,48 @@ func StatefulSetFromManifest(fileName, ns string) (*appsv1.StatefulSet, error) {
...
+
+	data, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to read html response body: %v", err)
+	}
+	return DsFromData(data)
+}
+
+// DsFromData reads a byte slice and returns the daemonset in it.
+func DsFromData(data []byte) (*appsv1.DaemonSet, error) {
```
Comment (oomichi) [inline test/e2e/framework/manifest/manifest.go:126]: If renaming the above function to `DaemonSetFromURL`, this function could be renamed to `DaemonSetFromData`.

**Review by oomichi — COMMENTED** [review]
Thanks for updating.

/lgtm

**Review by alejandrox1 — COMMENTED** [review]
Thank you @tanjunchen for working on this!

/approve
/milestone v1.19
